source/server/listener_manager_impl.h

```diff
@@ -205,6 +134,9 @@ class ListenerManagerImpl : public ListenerManager, Logger::Loggable<Logger::Id:
   ApiListenerOptRef apiListener() override;
 
   Quic::QuicStatNames& quicStatNames() { return quic_stat_names_; }
+  Filter::TcpListenerFilterConfigProviderManagerImpl& getTcpListenerConfigProviderManager() {
```

Comment:
If this a field that is expected to be in any ListenerManager it will probably be ok to define this method in the interface.

Reply:
Yeah, if we can add this interface in listener_manager.h, we can avoid the dynamic_cast below:
  Filter::ListenerFilterFactoriesList createListenerFilterFactoryList(
      const Protobuf::RepeatedPtrField<envoy::config::listener::v3::ListenerFilter>& filters,
      Configuration::ListenerFactoryContext& context) override {
    // TODO(yanjunxiang): cleanup this dynamic_cast.
    ListenerManagerImpl* listener_manager =
        dynamic_cast<ListenerManagerImpl*>(&(server_.listenerManager()));

However,

Adding this in the listener_manager.h is causing circular dependancy as below due to the return type is defined in config_discovery_impl.h:

ERROR: /usr/local/google/home/yanjunxiang/envoy-0507-tcp-only/envoy/source/common/config/BUILD:380:17: in cc_library rule //source/common/config:utility_lib: cycle in dependency graph:
    //source/exe:envoy-static (0c35d446f947aff734d4d198996f7fcc9a2be77ed6a6f1dd140eb1e298164c2d)
    //source/exe:envoy_main_entry_lib (0c35d446f947aff734d4d198996f7fcc9a2be77ed6a6f1dd140eb1e298164c2d)
    //source/exe:envoy_main_common_lib (0c35d446f947aff734d4d198996f7fcc9a2be77ed6a6f1dd140eb1e298164c2d)
    //source/exe:main_common_lib (0c35d446f947aff734d4d198996f7fcc9a2be77ed6a6f1dd140eb1e298164c2d)
    //source/server:hot_restart_lib (0c35d446f947aff734d4d198996f7fcc9a2be77ed6a6f1dd140eb1e298164c2d)
    //source/server:hot_restarting_parent (0c35d446f947aff734d4d198996f7fcc9a2be77ed6a6f1dd140eb1e298164c2d)
    //source/server:listener_manager_lib (0c35d446f947aff734d4d198996f7fcc9a2be77ed6a6f1dd140eb1e298164c2d)
.-> //source/common/config:utility_lib (0c35d446f947aff734d4d198996f7fcc9a2be77ed6a6f1dd140eb1e298164c2d)
|   //source/common/grpc:common_lib (0c35d446f947aff734d4d198996f7fcc9a2be77ed6a6f1dd140eb1e298164c2d)
|   //source/common/http:header_utility_lib (0c35d446f947aff734d4d198996f7fcc9a2be77ed6a6f1dd140eb1e298164c2d)
|   //source/common/http:utility_lib (0c35d446f947aff734d4d198996f7fcc9a2be77ed6a6f1dd140eb1e298164c2d)
|   //source/common/network:utility_lib (0c35d446f947aff734d4d198996f7fcc9a2be77ed6a6f1dd140eb1e298164c2d)
|   //source/common/network:default_socket_interface_lib (0c35d446f947aff734d4d198996f7fcc9a2be77ed6a6f1dd140eb1e298164c2d)
|   //source/common/event:dispatcher_includes (0c35d446f947aff734d4d198996f7fcc9a2be77ed6a6f1dd140eb1e298164c2d)
|   //source/common/signal:sigaction_lib (0c35d446f947aff734d4d198996f7fcc9a2be77ed6a6f1dd140eb1e298164c2d)
|   //source/common/signal:fatal_error_handler_lib (0c35d446f947aff734d4d198996f7fcc9a2be77ed6a6f1dd140eb1e298164c2d)
|   //source/common/signal:fatal_action_lib (0c35d446f947aff734d4d198996f7fcc9a2be77ed6a6f1dd140eb1e298164c2d)
|   //envoy/server:fatal_action_interface (0c35d446f947aff734d4d198996f7fcc9a2be77ed6a6f1dd140eb1e298164c2d)
|   //envoy/server:instance_interface (0c35d446f947aff734d4d198996f7fcc9a2be77ed6a6f1dd140eb1e298164c2d)
|   //envoy/server:listener_manager_interface (0c35d446f947aff734d4d198996f7fcc9a2be77ed6a6f1dd140eb1e298164c2d)
|   //source/common/filter:config_discovery_lib (0c35d446f947aff734d4d198996f7fcc9a2be77ed6a6f1dd140eb1e298164c2d)
`-- //source/common/config:utility_lib (0c35d446f947aff734d4d198996f7fcc9a2be77ed6a6f1dd140eb1e298164c2d)

We need to find a solution here.